dna/modify.go

```diff
@@ -1,139 +1,185 @@
+// start is closed, end is open, both are zero-based.
 func RangeToUpper(bases []Base, start int, end int) {
 	for i := start; i < end; i++ {
 		bases[i] = ToUpper(bases[i])
 	}
 }
 
-// start is closed, end is open, both are zero-based
+// RangeToLower changes the bases in a set range to lowercase.
+// start is closed, end is open, both are zero-based.
 func RangeToLower(bases []Base, start int, end int) {
 	for i := start; i < end; i++ {
 		bases[i] = ToLower(bases[i])
 	}
 }
 
+// AllToUpper changes all bases in a sequence to uppercase.
 func AllToUpper(bases []Base) {
 	RangeToUpper(bases, 0, len(bases))
 }
 
+// AllToLower changes all bases in a sequence to lowercase.
 func AllToLower(bases []Base) {
 	RangeToLower(bases, 0, len(bases))
 }
 
-func ComplementSingleBase(b Base) Base {
+// ComplementSingleBase returns the nucleotide complementary to the input base.
+func ComplementSingleBase(b Base) (Base, error) {
 	switch b {
 	case A:
-		return T
+		return T, nil
 	case C:
-		return G
+		return G, nil
 	case G:
-		return C
+		return C, nil
 	case T:
-		return A
+		return A, nil
 	case N:
-		return N
-	case a:
-		return t
-	case c:
-		return g
-	case g:
-		return c
-	case t:
-		return a
-	case n:
-		return n
+		return N, nil
+	case LowerA:
+		return LowerT, nil
+	case LowerC:
+		return LowerG, nil
+	case LowerG:
+		return LowerC, nil
+	case LowerT:
+		return LowerA, nil
+	case LowerN:
+		return LowerN, nil
 	case Gap:
-		return Gap
+		return Gap, nil
 	case Dot:
-		return Dot
+		return Dot, nil
+	case Nil:
+		return Nil, nil
 	default:
-		common.ExitIfError(fmt.Errorf("Error: trying to reverse ComplementSingleBase an unexpected base %d", b))
-		return N
+		return b, ErrUnrecognizedBase
 	}
 }
 
-func swap(alpha Base, beta Base) (Base, Base) {
-	return beta, alpha
+// ReverseComplement reverses a sequence of bases and complements each base.
+// Used to switch strands and maintain 5' -> 3' orientation.
+func ReverseComplement(bases []Base) error {
+	var err error
+	for i, j := 0, len(bases)-1; i <= j; i, j = i+1, j-1 {
+		bases[i], bases[j], err = complementSwap(bases[i], bases[j])
+		if err != nil {
+			break
+		}
+	}
+	return err
 }
 
-func ReverseComplement(bases []Base) {
-	for i, j := 0, len(bases)-1; i <= j; i, j = i+1, j-1 {
-		bases[i], bases[j] = swap(ComplementSingleBase(bases[i]), ComplementSingleBase(bases[j]))
+// Complement all bases in a sequence of bases.
+func Complement(bases []Base) error {
+	var err error
+	for i := range bases {
+		bases[i], err = ComplementSingleBase(bases[i])
+		if err != nil {
+			break
+		}
 	}
+	return err
 }
 
-func Complement(bases []Base) {
-	for i := 0; i < len(bases); i++ {
-		bases[i] = ComplementSingleBase(bases[i])
+// complementSwap complements the input bases and swaps their positions in the return.
+func complementSwap(alpha, beta Base) (Base, Base, error) {
+	var err error
+	alpha, err = ComplementSingleBase(alpha)
+	if err != nil {
+		return beta, alpha, err
```

Comment:
I think the functions that return errors will need  descriptions of what they return  upon errors.  This one will return the original bases/values swapped, but not complemented.